components/sup-protocol/protocols/ctl.proto

```diff
@@ -1,36 +1,37 @@
 syntax = "proto2";
+package sup.ctl;
 
 import "types.proto";
 
 // Networked progress bar for displaying a remote request's operation status over time.
 message NetProgress {
   // Number of total units until bar is complete.
-  optional uint64 total = 1;
+  required uint64 total = 1;
```

Comment:
So with prost it looks like we bring back 'required' to the protobuf definitions.. that seems we would be moving away from our internal guidance (which was based on Google's guidance for future proofing, ie required fields are considered harmful - https://github.com/google/protobuf/issues/2497)

Reply:
Prost actually doesn't care if the field is `optional` or `required`. I made this change intentionally because this is a type and not an RPC message. It's extremely unlikely that there won't ever be a case where a progress bar needs a value of total and, if it ever does, it'd be a totally different progress bar. So you'd instead implement a new protocol type for `MyCrazyProgressBar` and then have it contain whatever fields you need. Once you do that, you update the protocol messages which contain the progress bar to use your new crazy progress bar.

The conversation from this morning was me seeing if the changes I'm making here to cleanup the protocol a bit could make it into the release just so I didn't need to go through and audit the changes I made to make sure we won't have a breaking protocol change.

One more piece regarding required vs optional. I don't subscribe to the idea that "required is always bad" but instead "required is usually bad". Sometimes you *do* mean that something is required. You should use it when you know you will *never ever ever* change it from being required without throwing away the entire type. Working with types that have required fields is inherently easier than ones with optional fields and that is true in both the Prost and Protobuf crate. It signals to the developer that they don't ever need to worry about the case when a field will or will not be present.

Reply:
I like keeping in mind the distinction between types and messages when thinking about required vs. optional 👍